src/ln/channelmonitor.rs

```diff
@@ -1186,7 +1186,90 @@ impl ChannelMonitor {
 				}
 			}
 		} else {
-			//TODO: For each input check if its in our remote_commitment_txn_on_chain map!
+			let remote_commitment_txn_on_chain = self.remote_commitment_txn_on_chain.lock().unwrap();
+
+			let mut total_value = 0;
+			let mut inputs = Vec::new();
+			let mut key_materials = Vec::new();
+
+			for (txid, commitment_number) in remote_commitment_txn_on_chain.iter() {
+				if tx.input[0].previous_output.txid == *txid { // We're a HTLC-Timeout or HTLC-Success spending a remote commitment tx
+					let secret = self.get_secret(*commitment_number).unwrap();
+					let per_commitment_key = ignore_error!(SecretKey::from_slice(&self.secp_ctx, &secret));
+					let revocation_pubkey = match self.key_storage {
+						KeyStorage::PrivMode { ref revocation_base_key, .. } => {
+							let per_commitment_point = PublicKey::from_secret_key(&self.secp_ctx, &per_commitment_key);
+							ignore_error!(chan_utils::derive_public_revocation_key(&self.secp_ctx, &per_commitment_point, &PublicKey::from_secret_key(&self.secp_ctx, &revocation_base_key)))
+						},
+						KeyStorage::SigsMode { ref revocation_base_key, .. } => {
+							let per_commitment_point = PublicKey::from_secret_key(&self.secp_ctx, &per_commitment_key);
+							ignore_error!(chan_utils::derive_public_revocation_key(&self.secp_ctx, &per_commitment_point, &revocation_base_key))
+						},
+					};
+					let delayed_key = ignore_error!(chan_utils::derive_public_key(&self.secp_ctx, &PublicKey::from_secret_key(&self.secp_ctx, &per_commitment_key), &self.delayed_payment_base_key));
+					let redeemscript = chan_utils::get_revokeable_redeemscript(&revocation_pubkey, self.our_to_self_delay, &delayed_key);
+					let revokeable_p2wsh = redeemscript.to_v0_p2wsh();
+
+
+					if tx.output[0].script_pubkey == revokeable_p2wsh { //HTLC transactions have one txin, one txout
+						inputs.push(TxIn {
+							previous_output: BitcoinOutPoint {
+								txid: commitment_txid,
+								vout: 0,
+							},
+							script_sig: Script::new(),
+							sequence: 0xfffffffd,
+							witness: Vec::new(),
+						});
+						total_value += tx.output[0].value;
+					}
+
+					key_materials.push((redeemscript, per_commitment_key));
```

Comment:
Buffering key materials to create only one transaction even if inputs are from unrelated commitment_tx

Reply:
I'm not sure this is necessary - there cant be two inputs pointing to two different remote_commitment_txn_on_chain as the commitment txn would be double-spends.